ingest/lambdas/indexer.ts

```diff
@@ -0,0 +1,270 @@
+  }
+
+  // Index our document item.
+  await osClient.putDocumentItem(openSearchEndpoint, documentItem);
+};
+
+export interface ThumbnailRetrieveLink {
+  _thumbnailRetrieveLink: string
+}
+
+export const getThumbnailRetrieveLink = (
+  bitstreams: Bitstream[]
+): ThumbnailRetrieveLink | undefined => {
+  const thumbnailBitstream = findThumbnailBitstreamItem(bitstreams);
+  if (thumbnailBitstream) {
+    return {
+      _thumbnailRetrieveLink: thumbnailBitstream.retrieveLink,
+    };
+  }
+
+  return undefined;
+};
+
+export const getTerms = async (
+  openSearchEndpoint: string,
+  percolateFields: { title: string, contents: string }
+) => {
+  const terms = await osClient.percolateDocumentFields(
+    openSearchEndpoint,
+    percolateFields
+  );
+
+  return {
+    _terms: terms,
+  };
+};
+
+const explicitInvokeEventSchema = z.object({
+  uuid: z.string().uuid(),
+});
+
+const textExtractorSnsEventSchema = z.object({
+  Records: z.array(
+    z.object({
+      Sns: z.object({
+        Message: z.string().min(1), // This string should be an S3 Record.
+      }),
+    })
+  ).nonempty(),
+});
+
+const textExtractorS3RecordSchema = z.object({
+  Records: z.array(
+    z.object({
+      s3: z.object({
+        bucket: z.object({
+          name: z.string().min(1),
+        }),
+        object: z.object({
+          key: z.string().min(1),
+        }),
+      }),
+    })
+  ).nonempty(),
+});
+
+const isTextExtractorSnsEventSchema = zodTypeGuard(textExtractorSnsEventSchema);
+
+const isExplicitInvokeEventSchema = zodTypeGuard(explicitInvokeEventSchema);
+
+interface IngestRecord {
+  uuid: string
+  bitstreamTextKey?: string
+  bitstreamTextBucket?: string
+}
+
+const parseEvent = (event: unknown): IngestRecord[] => {
+  if (isTextExtractorSnsEventSchema(event)) {
+    return _.map(
+      event.Records,
+      (record) => {
+        const messageData = textExtractorS3RecordSchema.parse(
+          JSON.parse(record.Sns.Message)
+        );
+
+        const bitstreamTextBucket = messageData.Records[0].s3.bucket.name;
+        const bitstreamTextKey = messageData.Records[0].s3.object.key;
+        const [uuid] = bitstreamTextKey.split('.');
+
+        if (!uuid) {
+          throw new Error(`Ingest indexer event missing UUID ${JSON.stringify(event)}`);
+        }
+
+        return {
+          bitstreamTextBucket,
+          bitstreamTextKey,
+          uuid,
+        };
+      }
+    );
+  } if (isExplicitInvokeEventSchema(event)) {
+    return [{
+      uuid: event.uuid,
+    }];
+  }
+
+  throw new Error(`Ingest indexer invoked with unknown event: ${JSON.stringify(event)}`);
+};
+
+const index = async (
+  ingestRecord: IngestRecord,
+  dspaceItemBucket: string,
+  openSearchEndpoint: string
+): Promise<void> => {
+  // Get the DSpace item that starts this all off.
+  const dspaceItem = await getDSpaceItemFields(
+    dspaceItemBucket,
+    `${ingestRecord.uuid}.json`
+  );
+
+  // Add the PDF source if it exists.
```

Comment:
```suggestion
  // Get the PDF source if it exists.
```

Same for a couple other places in this function.